File: Review-06-Java-Shapes/Example-6/src/main/java/edu/odu/cs/cs330/examples/RunShapes.java
package edu.odu.cs.cs330.examples;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;

import java.util.List;
import java.util.ArrayList;

import java.util.Comparator;
import java.util.Iterator;

import edu.odu.cs.tkennedy.utilities.Utilities;
import static edu.odu.cs.tkennedy.utilities.Utilities.heading;
import static edu.odu.cs.tkennedy.utilities.Utilities.horizontalLine;
import static edu.odu.cs.tkennedy.utilities.Utilities.projectHeading;

// import edu.odu.cs.cs330.examples.shapes.*;
import edu.odu.cs.cs330.examples.shapes.Shape;
import edu.odu.cs.cs330.examples.shapes.ShapeFactory;

import edu.odu.cs.cs330.examples.shapes.io.ShapeIterator;

/**
 * This is the Java version of the previous C++ Shapes Inheritance Example.
 *
 * @author Thomas J Kennedy
 */
public class RunShapes {
    /**
     * This is the Program Title.
     */
    private static final String[] PROGRAM_HEADING = {
        "Objects & Inheritance: 2-D Shapes",
        "Thomas J. Kennedy"
    };

    /**
     * The default heading width.
     */
    private static final int H_WIDTH = 38;

    /**
     * This is the main function.
     *
     * @param args args[0] must specify the filename of a valid shape text file
     *
     * @throws CloneNotSupportedException if a Shape subclass can
     *     not be copied.
     */
    public static void main(String[] args)
        throws CloneNotSupportedException
    {
        BufferedReader shapesFile = null;
        try {
            // index is zero because java does command line arguments
            // the wrong way
            shapesFile = new BufferedReader(new FileReader(args[0]));
        }
        catch (ArrayIndexOutOfBoundsException e) {
            System.out.println("Usage: java -jar {jarfile} {inputTextFile}");
            System.exit(1);
        }
        catch (FileNotFoundException e) {
            System.out.println("File (" + args[0] + ") could not be opened.");
            System.exit(2);
        }

        // Print main program heading
        System.out.println(projectHeading(PROGRAM_HEADING, Utilities.W_WIDTH));

        //----------------------------------------------------------------------
        // Examine the ShapeFactory
        System.out.println(heading("Available Shapes", H_WIDTH, '*'));

        System.out.print(ShapeFactory.listKnown());
        System.out.println(horizontalLine('-', H_WIDTH));
        System.out.printf("%2d shapes available.%n", ShapeFactory.numberKnown());
        System.out.println();

        List<Shape> shapes = readShapes(shapesFile);

        // Print all the shapes
        System.out.println(heading("Display Shape Names", H_WIDTH, '*'));
        //printShapeNames(shapes);
        shapes.stream()
            .map(Shape::name)
            .forEach(System.out::println);

        // Using a stream
        System.out.println();
        System.out.println(heading("Display Shapes", H_WIDTH, '~'));
        shapes.stream()
            .forEach(System.out::println);

        System.out.println();

        System.out.println(heading("Display Largest Shape (Area)", H_WIDTH, '~'));
        Shape largestShape = shapes.parallelStream()
                           .max(Comparator.comparing(Shape::area))
                           .get();
        System.out.println(largestShape);

        System.out.println(heading("Display Smallest Shape (Perimeter)", H_WIDTH, '~'));
        Shape smallestShape = shapes.parallelStream()
                            .min(Comparator.comparing(Shape::perimeter))
                            .get();
        System.out.println(smallestShape);
    }

    /**
     * Read shapes from an input stream
     * and construct an `ArrayList<Shape>` object.
     *
     * @param buffer input source
     *
     * @return collection of read-in shapes
     *
     * @throws CloneNotSupportedException if the `ShapeFactory` fails to clone a
     *     model shape
     */
    private static List<Shape> readShapes(BufferedReader buffer)
        throws CloneNotSupportedException
    {
        List<Shape> collection = new ArrayList<Shape>();

        Iterator<Shape> it = new ShapeIterator(buffer);

        while (it.hasNext()) {
            Shape shp = it.next();

            if (shp != null) {
                collection.add(shp);
            }
        }

        return collection;
    }
}

File: Review-06-CPP-Shapes/Example-9/source/runShapes.cpp
// Thomas Kennedy
// Review Example: Inheritance and the Factory Model

#include <iostream>
#include <iomanip>
#include <fstream>
#include <vector>
#include <array>

#include <memory>
#include <iterator>
#include <algorithm>
#include <functional>

#include "utilities.h"
#include "shapeFactory.h"
// #include "ShapeCollection.h" // No longer needed

/** @file */

using namespace std;
using namespace utilities;

constexpr std::array<string_view, 2> PROGRAM_HEADING = {
    "Objects & Inheritance: 2-D Shapes",
    "Thomas J. Kennedy"
};  ///< Program Title

using ShapeCollection = std::vector<std::unique_ptr<Shape>>;

/**
 * Read shapes from an input stream
 * and construct a `ShapeCollection` object
 */
ShapeCollection readShapes(std::istream& ins);

/**
 * Print shapes from a `ShapeCollection` to a
 * specified output stream
 */
// void printShapes(std::ostream& outs, const ShapeCollection& toPrint);
ostream& operator<<(std::ostream& outs, const ShapeCollection& toPrint);

/**
 * Print shape names for all `Shape`s in a `ShapeCollection` to a
 * specified output stream
 */
void printShapeNames(std::ostream& outs, const ShapeCollection& toPrint);

/**
 * Implementation of a `transform_if` based on `transform` as documented at
 * https://en.cppreference.com/w/cpp/algorithm/transform
 *
 * @tparam InputIt iterator for the source container
 * @tparam OutputIt iterator that allows insertion into an output container
 * @tparam UnaryOp1 boolean function that determines whether include or skip
 *         source data
 * @tparam UnaryOp2 function that transforms data
 *
 * @todo discuss why this is incomplete
 */
template<class InputIt, class OutputIt, class UnaryOp1, class UnaryOp2>
OutputIt transform_if(InputIt itInputStart,
                      InputIt itInputEnd,
                      OutputIt itOutput,
                      UnaryOp1 conditionOperation,
                      UnaryOp2 transformOperation)
{
    while (itInputStart != itInputEnd) {

        if (conditionOperation(*itInputStart)) {
            *itOutput++ = transformOperation(*itInputStart);
        }

        // itInputStart++;
        ++itInputStart;  // Pre-increment avoids a temporary copy.
    }

    return itOutput;
}

/**
 * This program accepts command line
 * arguments
 */
int main(int argc, char** argv)
{
    if (argc < 2) {
        cout << "No input file provided." << "\n";
        cout << "Usage: " << argv[0] << " input_file" << "\n";
        return 1;
    }

    ifstream shapesFile(argv[1]);

    if (!shapesFile) {
        cout << "Error: " << argv[1] << "could not be opened" << "\n";
        return 2;
    }

    // Set formatting
    cout.precision(4);
    cout.setf(ios::fixed);

    cout << generateProjectHeading(PROGRAM_HEADING);

    // Examine the ShapeFactory & list available shapes
    cout << generateHeading<'~', 38>("Available Shapes");

    ShapeFactory::listKnown(cout);
    cout << generateHorizontalLine<'-', 38>();
    cout << right << setw(2)
         << ShapeFactory::numberKnown()
         << " shapes available."
         << "\n";

    cout << "\n";

    // Create 5 "Random" Shapes
    ShapeCollection shapes = readShapes(shapesFile);

    // Print all the shapes
    cout << generateHeading<'~', 38>("Display All Shapes");
    // printShapes(cout, shapes);
    cout << shapes;

    cout << generateHeading<'~', 38>("Display Shape Names");
    // printShapeNames(cout, shapes);  // Should I keep this?
    std::transform(shapes.begin(), shapes.end(),
                   std::ostream_iterator<std::string>(cout, "\n"),
                   [](const unique_ptr<Shape>& shp) -> std::string {
                       return shp->name();
                   });

    //--------------------------------------------------------------------------
    cout << generateHeading<'~', 38>("Display Largest Shape (Area)");
    ShapeCollection::const_iterator it;

    it = std::max_element(shapes.begin(), shapes.end(),
                          [](const auto& lhs, const auto& rhs) {
                              return (lhs)->area() < (rhs)->area();
                          });

    // const unique_ptr<Shape>& data = *it;
    // cout << *data << "\n";
    cout << *(*it) << "\n";

    //--------------------------------------------------------------------------
    cout << generateHeading<'~', 38>("Display Smallest Shape (Perimeter)");
    it = std::min_element(shapes.begin(), shapes.end(),
                          [](const auto& lhs, const auto& rhs) {
                              return (lhs)->perimeter() < (rhs)->perimeter();
                          });
    cout << *(*it) << "\n";

    //--------------------------------------------------------------------------
    cout << generateHeading<'~', 38>("Display Sorted by Name");
    std::sort(shapes.begin(), shapes.end(),
              [](const auto& lhs, const auto& rhs) {
                  return (lhs)->name() < (rhs)->name();
              });

    // printShapes(cout, shapes);
    cout << shapes;

    return 0;
}

//------------------------------------------------------------------------------
ShapeCollection readShapes(std::istream& ins)
{
    ShapeCollection collection;

    std::istream_iterator<Shape*> it(ins);
    std::istream_iterator<Shape*> end;

    // Let us re-enact the Back-to-the-Future I Guitar Scene!
    //
    // std::copy_if would copy raw pointers, not unique pointers.
    //
    // To get our desired result we would need a copy-if folowed by a transform.
    //
    /*
    for_each(it, end,
             [&collection](Shape* s) {
                  if (s != nullptr) {
                      collection.emplace_back(s);
                  }
             });
    */

    transform_if(it, end,
                 std::back_inserter(collection),
                 [](Shape* rawPtr) -> bool {
                     return rawPtr != nullptr;
                 },
                 [](Shape* rawPtr) -> std::unique_ptr<Shape> {
                     return std::unique_ptr<Shape>(rawPtr);
                 });

    return collection;
}

//------------------------------------------------------------------------------
inline
std::ostream& operator<<(std::ostream& outs, const unique_ptr<Shape>& prt)
{
    // prt->display(outs);
    outs << *prt;

    return outs;
}

//------------------------------------------------------------------------------
// void printShapes(std::ostream& outs, const ShapeCollection& toPrint)
ostream& operator<<(std::ostream& outs, const ShapeCollection& toPrint)
{
    /*
    for (const std::unique_ptr<Shape>& s : toPrint) {
        // outs << s << "\n";  // oops
        outs << *s << "\n";
    }
    */

    std::copy(toPrint.begin(), toPrint.end(),
              std::ostream_iterator<unique_ptr<Shape>>(outs, "\n"));

    return outs;
}

//------------------------------------------------------------------------------
void printShapeNames(std::ostream& outs, const ShapeCollection& toPrint)
{
    /*
    for (const std::unique_ptr<Shape>& s : toPrint) {
        outs << s->name() << "\n";
    }
    */

    std::transform(toPrint.begin(), toPrint.end(),
                   std::ostream_iterator<std::string>(outs, "\n"),
                   [](const std::unique_ptr<Shape>& s) -> std::string {
                       return s->name();
                   });
}
